trieste/acquisition/function.py

```diff
@@ -509,6 +512,155 @@ def prepare_acquisition_function(
+    generalized expected improvements for easier math operation, corresponding to Psi
+    function and nu function calculations.
+    Mainly referring Eq. 44, and Eq. 45
+    Note:
+    1. Since in Trieste we do not assume the use of a certain non-dominated partition algorithm,
+       we do not assume the last dimension partitioned cell has only one (lower) bound
+       (which is used in the :cite:`yang2019efficient` paper). This is not as efficient as
+        the original paper, but is applicable to different non-dominated partition algorithm.
+    2. The Psi and nu function in the original paper is defined for a maximization problem. To
+       make use of the same notation for easier following, we inverse our problem (as maximization)
+       to make use of the same equation.
+
+    :param model: The model of the objective function.
+    :param pareto: Pareto class
+    :param reference_point The reference point for calculating hypervolume
+    :return The expected_hv_improvement acquisition function modified for objective
+        minimisation. This function will raise :exc:`ValueError` or
+        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+    """
+
+    def acquisition(x: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(x, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        normal = tfp.distributions.Normal(
+            loc=tf.zeros(shape=1, dtype=x.dtype), scale=tf.ones(shape=1, dtype=x.dtype)
+        )
+
+        def Psi(a, b, mean, std) -> TensorType:
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return std * normal.prob((b - mean) / std) + (mean - a) * (
+                1 - normal.cdf((b - mean) / std)
+            )
+
+        def nu(lb, ub, mean, std) -> TensorType:
+            """
+            A special expected improvement given fixed improvement,
+            defined at Eq. 25 in :cite:`yang2019efficient`
+            :param lb: shape [..., num_cells, out_dim] slice lower bound
+            :param ub: shape [..., num_cells, out_dim] slice upper bound
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return (ub - lb) * (1 - normal.cdf((ub - mean) / std))
+
+        candidate_mean, candidate_var = model.predict(tf.squeeze(x, -2))
+        candidate_std = tf.sqrt(candidate_var)
+
+        # calc ehvi assuming maximization
```

Comment:
In trieste, we have tried to not have any comment lines as they can get separated from the code they refer to. Try to keep discussion at the start of the function (where we discuss the input/outputs) or inline (like you did with shapes below). This comment is unnecessary anyway,